include/ginkgo/core/base/executor.hpp

```diff
@@ -1450,20 +1455,26 @@ class CudaExecutor : public detail::ExecutorBase<CudaExecutor>,
 
     static void increase_num_execs(unsigned device_id)
     {
-        std::lock_guard<std::mutex> guard(mutex[device_id]);
-        num_execs[device_id]++;
+#ifdef GINKGO_BUILD_CUDA
+        // increase the Cuda Device count only when ginkgo build cuda
+        std::lock_guard<std::mutex> guard(device_class::get_mutex(device_id));
+        device_class::get_num_execs(device_id)++;
+#endif  // GINKGO_BUILD_CUDA
```

Comment:
This strongly links the public interface with the implementation, which might break our assumptions about being able to just swap out shared library stubs for actual implementations. Can we move this function implementation into `core`?

Reply:
I see, but is it okay for the device_class choice?

Reply:
Can you elaborate on that? In the current setting, we may violate the ODR rule by providing different implementations for the function in a Ginkgo build with or without CUDA. It would be great if we could keep the public interface entirely agnostic of which backends we built.

Reply:
I use GINKGO_HIP_PLATFORM_NVCC to choose the device_class.
does it destroy the rule?

Reply:
That's not necessarily an ODR issue, since it is only related to friend declarations. But still, it would be great if we could have a mostly consistent interface across compilers or configurations

Reply:
it also lead the issue when we swap the shared_memory, right?

Reply:
you mean the unified memory parameter for executors?

Reply:
Sorry, I mixed two pull request...
it also lead the issue when we swap the shared library, right?

Reply:
@upsj I move all device_class related function to cpp and always make NvidiaDevice have friend class Cuda/Hip and AmdDevice have Hip. Only select the device_class in cpp not in class.